runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,431 @@
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.BiFunction;
+import java.util.function.Predicate;
+
+/**
+ * TxnContext is the access layer for binding all the CorfuStore CRUD operations.
+ * It can help reduce the footprint of a CorfuStore transaction by only having writes in it.
+ * All mutations/writes are aggregated and applied at once a the time of commit() call
+ * where a real corfu transaction is started.
+ *
+ * Created by hisundar, @wenbinzhu, @pankti-m on 2020-09-15
+ */
+@Slf4j
+public class TxnContext implements AutoCloseable {
+
+    private final ObjectsView objectsView;
+    private final TableRegistry tableRegistry;
+    private final String namespace;
+    private final IsolationLevel isolationLevel;
+    private final List<Runnable> operations;
+    private final Set<Table> tablesInTxn;
+    private long txnStartTime = 0L;
+    private static final byte READ_ONLY  = 0x01;
+    private static final byte WRITE_ONLY = 0x02;
+    private static final byte READ_WRITE = 0x03;
+    private byte txnType;
+
+    /**
+     * Creates a new TxnContext.
+     *
+     * @param objectsView   ObjectsView from the Corfu client.
+     * @param tableRegistry Table Registry.
+     * @param namespace     Namespace boundary defined for the transaction.
+     * @param isolationLevel How should this transaction be applied/evaluated.
+     */
+    @Nonnull
+    TxnContext(@Nonnull final ObjectsView objectsView,
+               @Nonnull final TableRegistry tableRegistry,
+               @Nonnull final String namespace,
+               @Nonnull final IsolationLevel isolationLevel) {
+        this.objectsView = objectsView;
+        this.tableRegistry = tableRegistry;
+        this.namespace = namespace;
+        this.isolationLevel = isolationLevel;
+        this.operations = new ArrayList<>();
+        this.tablesInTxn = new HashSet<>();
+        txBeginInternal(); // May throw exception if transaction was already started
+    }
+
+    public  <K extends Message, V extends Message, M extends Message>
+    Table<K, V, M> getTable(@Nonnull final String tableName) {
+        return this.tableRegistry.getTable(this.namespace, tableName);
+    }
+
+    /**
+     *************************** WRITE APIs *****************************************
+     */
+
+    /**
+     * All write api must be validate to ensure that the table belongs to the namespace.
+     *
+     * @param table - table being written to
+     * @param <K> -type of the key
+     * @param <V> - type of the payload/value
+     * @param <M> - type of the metadata
+     */
+    private <K extends Message, V extends Message, M extends Message>
+    void validateTableWrittenIsInNamespace(@Nonnull Table<K, V, M> table) {
+        if (!table.getNamespace().equals(namespace)) {
+            throw new IllegalArgumentException("TxnContext can't apply table from namespace "
+                    + table.getNamespace() + " to transaction on namespace " + namespace);
+        }
```

Comment:
Why is this validation only for write operations? Should this also be done for reads?

Reply:
reads can query any namespace - we want to retain this behavior to not break any existing use cases.

Reply:
Initially reads can query an namespace because reads are outside of txnContext,  but now reads are part of txnContext. Letting reads to query any namespace,  but writes to query only one namespace seems a bit weird...